src/main/java/com/todo/backend/Controller.java

```diff
@@ -1,54 +1,302 @@
+        if (isSame && lastResponse != null) {
+            filtered = lastResponse;
+        } else {
+            // DIFFERENT PARAMS OR THERE'S NOT A PREVIOUS
+            filtered = new ArrayList<>(DB);
+
+            // FILTER (IF ANY)
+            if (done != null || text != null || priority != null) {
+                filtered = this.filter(done, text, priority);
+            }
+
+            // SORT BY (IF ANY)
+            if (sortDueDate != null || sortPriority != null) {
+                sortList(filtered, sortPriority, sortDueDate);
+            }
+
+            lastResponse = filtered; // SAVE THE LIST FOR NEXT
+        }
+
+        // EVEN IF WE GOT THE SAME ANSWER, WE CAN SEND A DIFFERENT PAGE TO CLIENT
+        int maxpage = (int) Math.max(Math.ceil(filtered.size() / 10f), 1);
+        int page = (int) Math.min(Math.max(pag, 1), maxpage);
+        List<ToDo> slice = filtered.subList(10 * (page - 1), Math.min(10 * page, filtered.size()));
+
+        return (new GETResponse(slice, page, maxpage, lm));
+    }
+
+    @RequestMapping(method = { RequestMethod.GET }, value = { "/todos/{id}" })
+    public ResponseEntity<?> getToDo(@PathVariable int id) {
+        if (id < 0)
+            return new ResponseEntity<>("Invalid ID", HttpStatus.BAD_REQUEST);
+
+        for (ToDo toDo : DB) {
+            if (toDo.getId() == id)
+                return new ResponseEntity<ToDo>(toDo, HttpStatus.FOUND);
+        }
+
+        return new ResponseEntity<>("Not Found", HttpStatus.NOT_FOUND);
+    }
+    // #endregion
+
+    // #region ################################ POST
+    @RequestMapping(method = { RequestMethod.POST }, value = { "/todos" })
+    public ResponseEntity<String> createToDo(@RequestBody BasicToDo toDo) {
+        if (toDo.getText().length() <= 0 || toDo.getText().length() > 120)
+            return new ResponseEntity<>("Text length must be 120 or lower", HttpStatus.BAD_REQUEST);
+
+        boolean stat = DB.add(
+                new ToDo(
+                        toDo.getText(),
+                        toDo.getPriority(),
+                        toDo.getDue_date()));
+
+        if (!stat)
+            return new ResponseEntity<>("Server error", HttpStatus.INTERNAL_SERVER_ERROR);
+
+        this.metrics.needsRecalculate();
+        this.lastResponse = null; // FORCE RECALCULATE
+        return new ResponseEntity<>("Added new ToDo item", HttpStatus.CREATED);
+    }
+
+    @RequestMapping(method = { RequestMethod.PUT }, value = { "/todos/{id}" })
+    public ResponseEntity<String> updateToDo(@PathVariable int id, @RequestBody Map<String, String> body)
+            throws ParseException {
+        if (id < 0)
+            return new ResponseEntity<>("Invalid ID", HttpStatus.BAD_REQUEST);
+
+        for (int i = 0; i < DB.size(); i++) {
+            ToDo todo = DB.get(i);
+            if (todo.getId() == id) {
+
+                String text = body.get("text");
+                if (text != null)
+                    todo.setText(text);
```

Comment:
notice how the validation is done here, so there's no point to have it inside the ToDo setter